clients/python/src/osparc/_utils.py

```diff
@@ -46,38 +48,40 @@ def __init__(
         self._client: httpx.Client = httpx.Client(
             auth=auth, base_url=base_url, follow_redirects=True
         )
+        self._page: Optional[Page] = None
+        self._page_item_counter = 0
 
     def __del__(self):
         self._client.close()
 
+    def __next__(self):
+        if self._page is None:
+            self._page = self._first_page_callback()
+        if self._page_item_counter > (len(self._page.items) - 1):
+            next_page_url = self._page.links.next
+            if next_page_url is None:
+                self._page = None
+                self._page_item_counter = 0
+                raise StopIteration
+            response = self._client.get(next_page_url)
+            self._page = self._api_client._ApiClient__deserialize(
+                response.json(), type(self._page)
+            )
+            self._page_item_counter = 0
+        next_item = self._page.items[self._page_item_counter]
+        self._page_item_counter += 1
+        return next_item
+
     def __len__(self) -> int:
         """Number of elements which the iterator can produce"""
         page: Page = self._first_page_callback()
         assert isinstance(page.total, int)
         return page.total
 
-    def __iter__(self) -> Generator[T, None, None]:
-        """Returns the generator"""
-        if len(self) == 0:
-            return
-        page: Page = self._first_page_callback()
-        while True:
-            assert page.items is not None
-            assert isinstance(page.total, int)
-            yield from page.items
-            if page.links.next is None:
-                return
-            response: httpx.Response = self._client.get(page.links.next)
-            try:
-                response.raise_for_status()
-            except httpx.HTTPStatusError as e:
-                raise RequestError(f"{e}") from e
-            page = self._api_client._ApiClient__deserialize(response.json(), type(page))
-
 
 async def file_chunk_generator(
     file: Path, chunk_size: int
-) -> AsyncGenerator[Tuple[bytes, int], None]:
+) -> AsyncGenerator[Tuple[bytes, int, bool], None]:
```

Comment:
What about a `NamedTuple` in the return with fields `chunk`, `nbytes`, `done`?

Reply:
Done